### ISSUE TITLE
Train model for MS lesion segmentation on MP2RAGE data

### ISSUE THREAD

**Issue by jcohenadad** [ivadomed/ivadomed#821]
### Context

Researchers from University of Basel shared MP2RAGE data from 30 subjects with MS. Data include manual segmentation of MS lesions.

Data are under: `data.neuro.polymtl.ca:basel-mp2rage`

### Goal

Train a deep learning model to segment MS lesions in the spinal cord
Make that model available in SCT.

**Comment by uzaymacar** [comment]
For reproducibility and tracking progress purposes, below is the `ivadomed` config file I am running for this dataset. The model I'm using is `Modified3DUNet`.

```
{
    "command": "train",
    "gpu_ids": [0],
    "path_output": "basel_lesionseg_output",
    "model_name": "basel_lesionseg_model",
    "debugging": true,
    "object_detection_params": {
        "object_detection_path": null,
        "safety_factor": [1.0, 1.0, 1.0]
    },
    "loader_parameters": {
        "path_data": ["basel-mp2rage"],
        "subject_selection": {
            "n": [],
            "metadata": [],
            "value": []
        },
        "target_suffix": ["_lesion-manual"],
        "extensions": [],
        "roi_params": {
            "suffix": null,
            "slice_filter_roi": null
        },
        "contrast_params": {
            "training_validation": ["UNIT1"],
            "testing": ["UNIT1"],
            "balance": {}
        },
        "slice_filter_params": {
            "filter_empty_mask": true,
            "filter_empty_input": true
        },
        "slice_axis": "axial",
        "multichannel": false,
        "soft_gt": false
    },
    "split_dataset": {
        "fname_split": null,
        "random_seed": 5,
        "center_test": [],
        "method": "per_patient",
        "balance": null,
        "train_fraction": 0.75,
        "test_fraction": 0.15
    },
    "training_parameters": {
        "batch_size": 4,
        "loss": {
            "name": "DiceLoss"
        },
        "training_time": {
            "num_epochs": 500,
            "early_stopping_patience": 50,
            "early_stopping_epsilon": 0.001
        },
        "scheduler": {
            "initial_lr": 1e-4,
            "lr_scheduler": {
                "name": "CosineAnnealingLR",
                "base_lr": 1e-5,
                "max_lr": 1e-4
            }
        },
        "balance_samples": {
            "applied": false,
            "type": "gt"
        },
        "mixup_alpha": null,
        "transfer_learning": {
            "retrain_model": null,
            "retrain_fraction": 1.0,
            "reset": true
        }
    },
    "default_model": {
        "name": "Unet",
        "dropout_rate": 0.3,
        "bn_momentum": 0.9,
        "depth": 3,
        "is_2d": true,
        "final_activation": "sigmoid"
    },
    "uncertainty": {
        "epistemic": false,
        "aleatoric": false,
        "n_it": 0
    },
    "postprocessing": {"binarize_prediction": {"thr": 0.5}},
    "evaluation_parameters": {},
    "transformation": {
        "Resample": {
            "wspace": 1.0,
            "hspace": 1.0,
            "dspace": 1.0
        },
        "CenterCrop": {
            "size": [128, 128, 128]
        },
        "RandomAffine": {
            "degrees": 10,
            "scale": [0.1, 0.1, 0.1],
            "translate": [0.1, 0.1, 0.1],
            "applied_to": ["im", "gt"],
            "dataset_type": ["training"]
        },
        "ElasticTransform": {
            "alpha_range": [25.0, 35.0],
            "sigma_range": [3.5, 5.5],
            "p": 0.3,
            "applied_to": ["im", "gt"],
            "dataset_type": ["training"]
        },
        "NumpyToTensor": {},
        "NormalizeInstance": {
            "applied_to": ["im"]
        }
    },
    "Modified3DUNet": {
        "applied": true,
        "length_3D": [128, 128, 128],
        "stride_3D": [64, 64, 64],
        "attention": false,
        "n_filters": 8
    }
}
```

Currently, the best validation score I reach with this config is `0.2032`. Next steps include:
* Playing around with batch size and learning rate scheduler.
* Playing around with different transformations (we could add gamma contrast to `ivadomed`, which was helpful in the MSSEG-2 challenge).

**Comment by uzaymacar** [comment]
Based on previous discussions, I used the following preprocessing script with SCT which (i) extracts the spinal-cord and (ii) crops the VOI using the bounding box given by the spinal-cord mask for each of the 30 patients:

```
#!/bin/bash
#
# Preprocess data.
#
# Dependencies:
# - FSL <TODO: VERSION>
# - SCT <TODO: VERSION>
#
# Usage:
#   ./preprocess_data.sh <SUBJECT>
#
#
# Manual segmentations or labels should be located under:
# PATH_DATA/derivatives/labels/SUBJECT/<CONTRAST>/

# The following global variables are retrieved from the caller sct_run_batch
# but could be overwritten by uncommenting the lines below:
# PATH_DATA_PROCESSED="~/data_processed"
# PATH_RESULTS="~/results"
# PATH_LOG="~/log"
# PATH_QC="~/qc"

# Uncomment for full verbose
set -x

# Immediately exit if error
set -e -o pipefail

# Exit if user presses CTRL+C (Linux) or CMD+C (OSX)
trap "echo Caught Keyboard Interrupt within script. Exiting now.; exit" INT

# Retrieve input params and other params
SUBJECT=$1

# get starting time:
start=`date +%s`


# SCRIPT STARTS HERE
# ==============================================================================
# Display useful info for the log, such as SCT version, RAM and CPU cores available
sct_check_dependencies -short

# Go to folder where data will be copied and processed
cd $PATH_DATA_PROCESSED

# Copy BIDS-required files to processed data folder (e.g. list of participants)
if [[ ! -f "participants.tsv" ]]; then
  rsync -avzh $PATH_DATA/participants.tsv .
fi
if [[ ! -f "participants.json" ]]; then
  rsync -avzh $PATH_DATA/participants.json .
fi
if [[ ! -f "dataset_description.json" ]]; then
  rsync -avzh $PATH_DATA/dataset_description.json .
fi
if [[ ! -f "README" ]]; then
  rsync -avzh $PATH_DATA/README .
fi

# Copy source images
rsync -avzh $PATH_DATA/$SUBJECT .

# Copy segmentation GTs
mkdir -p derivatives/labels
rsync -avzh $PATH_DATA/derivatives/labels/$SUBJECT derivatives/labels/.

# (1) Go to subject folder for source images
cd ${SUBJECT}

# TODO: re-think how file variable is defined-- not clean to have folders in there
file_onlyfile="${SUBJECT}_UNIT1"
file="anat/${SUBJECT}_UNIT1"

# Spinal cord extraction
sct_deepseg_sc -i ${file}.nii.gz -c t1 -o ${file_onlyfile}_seg.nii.gz

# Dilate spinal cord mask
sct_maths -i ${file_onlyfile}_seg.nii.gz -dilate 5 -shape ball -o ${file_onlyfile}_seg_dilate.nii.gz

# Compute the bounding box coordinates of SC mask for cropping the VOI
# NOTE: `fslstats -w returns the smallest ROI <xmin> <xsize> <ymin> <ysize> <zmin> <zsize> <tmin> <tsize> containing nonzero voxels
bbox_coords=$(fslstats "${file_onlyfile}"_seg_dilate.nii.gz -w)

# Apply the SC mask to the final forms of both sessions
fslmaths ${file}.nii.gz -mas ${file_onlyfile}_seg_dilate.nii.gz ${file_onlyfile}_masked.nii.gz

# Crop the VOI based on SC mask to minimize the input image size
fslroi ${file_onlyfile}_masked.nii.gz ${file_onlyfile}_masked.nii.gz $bbox_coords

# (2) Go to subject folder for segmentation GTs
cd $PATH_DATA_PROCESSED/derivatives/labels/$SUBJECT

file_gt_onlyfile="${SUBJECT}_UNIT1_lesion-manual"
file_gt="anat/${SUBJECT}_UNIT1_lesion-manual"

# Apply the SC mask to the final forms of all segmentation GTs
fslmaths ${file_gt}.nii.gz -mas $PATH_DATA_PROCESSED/$SUBJECT/${file_onlyfile}_seg_dilate.nii.gz ${file_gt_onlyfile}_masked.nii.gz

# Crop the VOI based on SC mask to minimize the GT image size
fslroi ${file_gt_onlyfile}_masked.nii.gz ${file_gt_onlyfile}_masked.nii.gz $bbox_coords

# Go back to parent folder (i.e. get ready for next subject call!)
cd $PATH_DATA_PROCESSED

# Display useful info for the log
end=`date +%s`
runtime=$((end-start))
echo
echo "~~~"
echo "SCT version: `sct_version`"
echo "Ran on:      `uname -nsr`"
echo "Duration:    $(($runtime / 3600))hrs $((($runtime / 60) % 60))min $(($runtime % 60))sec"
echo "~~~"
```

As a result, we were left with far smaller volumes which will hopefully make training the model simpler. As a result of the preprocessing, for `basel-mp2rage` we now have 30 volumes each with a resolution of `1mm x 1mm x 1mm` and with the following sizes:

```
SIZES:  Counter({(22, 95, 23): 1, (26, 99, 25): 1, (24, 98, 25): 1, (30, 88, 27): 1, (42, 108, 29): 1, (34, 102, 25): 1, (34, 111, 22): 1, (33, 115, 25): 1, (28, 101, 29): 1, (29, 94, 27): 1, (30, 100, 23): 1, (33, 92, 26): 1, (28, 103, 22): 1, (23, 100, 22): 1, (28, 78, 26): 1, (38, 96, 28): 1, (28, 86, 23): 1, (30, 99, 26): 1, (23, 109, 31): 1, (28, 101, 25): 1, (32, 101, 28): 1, (24, 106, 27): 1, (26, 104, 24): 1, (31, 86, 28): 1, (38, 90, 25): 1, (25, 90, 23): 1, (25, 80, 26): 1, (23, 98, 25): 1, (21, 102, 23): 1, (23, 96, 23): 1})
```

Based on these sizes, I am using a center-crop size of `(64, 128, 32)`. I have the updated the config file in my first message with the current parameters I'm using.

**Comment by uzaymacar** [comment]
Just to update this issue, we are developing the models in a new ivadomed repository: https://github.com/ivadomed/model_seg_ms_mp2rage.

**Comment by jcohenadad** [comment]
completed
